### ISSUE TITLE
Reject null-suppression operator in patterns

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#53484]
Branch: extended-property-patterns

We've relaxed pattern parsing to accept null-suppression and pointer access for error recovery. 
As a consequence, the following is now valid:

```cs
public class C {
    public void M() {
        const C c = null;
        if (this is c!) {}       
    }
}
```
Any other case is caught by "constant expected" error, however, that specific scenario satisfies all the other requirements.

**Comment by jcouv** [comment]
> We've relaxed pattern parsing to accept null-suppression and pointer access for error recovery.

Thanks for identifying this regression.
Error recovery should mean "keep error scenarios as errors, but better ones". I didn't realize those scenarios became accepted.
We should definitely still disallow.

Link to test plan: https://github.com/dotnet/roslyn/issues/52468

**Comment by RikkiGibson** [comment]
It wasn't clear to me why nullable suppression shouldn't be allowed in a constant pattern. Could you please clarify?

**Comment by jcouv** [comment]
@RikkiGibson A suppression would only be useful on a constant pattern if we ever produced nullability warnings there.

**Comment by jcouv** [comment]
Fixed by https://github.com/dotnet/roslyn/pull/53594 in feature branch (planned to merge for 17.0)
